index.js

```diff
@@ -195,15 +195,44 @@ DecryptionStream.prototype._transform = function(chunk, enc, cb) {
 };
 
 DecryptionStream.prototype._flush = function(cb) {
-  var data = Buffer.concat(this._cipherTextChunks);// this could be rewritten to avoid doing this
-  var mac = data.slice(-16);
+  var mac = pullOutMac(this._cipherTextChunks);
+  if (!mac) {
+    return this.emit('error', new Error('insufficient data'));
+  }
   this._decipher.setAuthTag(mac);
-  var decrypted = this._decipher.update(data.slice(0, -16));
+  var decrypted = this._cipherTextChunks.map(function(item) {
+    return this._decipher.update(item);
+  }, this);
   try {
     this._decipher.final();
-  } catch(e) {
+  } catch (e) {
     return cb(e);
   }
-  this.push(decrypted);
+  decrypted.forEach(function(item) {
+    this.push(item);
+  }, this);
   cb();
 };
+
+function pullOutMac(array) {
+  var macBits = [];
+  var macSoFar = 0;
+  var current, overflow;
+  while (macSoFar !== GCM_MAC_LENGTH && array.length) {
+    current = array.pop();
+    if (macSoFar + current.length <= GCM_MAC_LENGTH) {
+      macBits.push(current);
+      macSoFar += current.length;
+    } else {
+      overflow = (macSoFar + current.length) - GCM_MAC_LENGTH;
```

Comment:
Could: rename to `macStartIndex` or similar. Overflow is a slightly confusing name when used in context below.